File: problems/linked-list/0739-daily-temperatures.js
/**
 * #739 Daily Temperatures
 * Pattern: monotonic decreasing stack of unresolved day indexes.
 * When a warmer day arrives, it resolves each cooler index at the top.
 * Time: O(n) | Space: O(n)
 *
 * @param {number[]} temperatures
 * @return {number[]}
 */
function dailyTemperatures(temperatures) {
  const waits = Array(temperatures.length).fill(0)
  const unresolved = []

  for (let day = 0; day < temperatures.length; day += 1) {
    while (
      unresolved.length > 0 &&
      temperatures[day] > temperatures[unresolved[unresolved.length - 1]]
    ) {
      const earlierDay = unresolved.pop()
      waits[earlierDay] = day - earlierDay
    }
    unresolved.push(day)
  }

  return waits
}

module.exports = dailyTemperatures

File: problems/linked-list/0739-daily-temperatures.py
"""#739 Daily Temperatures
Keep unresolved day indexes in a monotonic decreasing-temperature stack.
Time: O(n) | Space: O(n)
"""


class Solution:
    def dailyTemperatures(self, temperatures):
        waits = [0] * len(temperatures)
        unresolved = []

        for day, temperature in enumerate(temperatures):
            while (
                unresolved
                and temperature > temperatures[unresolved[-1]]
            ):
                earlier_day = unresolved.pop()
                waits[earlier_day] = day - earlier_day
            unresolved.append(day)

        return waits
